rfd/00XX-proxy-peering.md

```diff
@@ -0,0 +1,222 @@
+int64 NonceID = 14 [ (gogoproto.jsontag) = "nonce_id,omitempty" ];
+```
+
+Since each proxy already keeps a cache of `Servers` there will be no additional mechanism required to replicate this information.
+
+Each agent will be responsible for updating the `ProxyID` as it connects and reconnects to proxy servers. This will be done over the existing periodic heartbeats to the auth server. If `proxy_peering` is not enabled in the `ClusterNetworkingConfig` the `ProxyID` should not be included.
+
+The `Nonce` will start at 0 and be incremented with each update sent to the auth server. On each restart of the teleport agent a new `NonceID` will be randomly generated. The auth server will reject any updates where the `heartbeat.nonce < previous.nonce && heartbeat.nonce_id == previous.nonce_id`.
+
+### API Clients
+Each proxy will need to manage multiple grpc clients, one for each peer proxy. Client connections will be created as peer proxies are discovered. Similar to the agent pools current behavior, clients can be expired if the connection fails and the peer hasn't heartbeated to the backend for a certain amount of time.
+
+Transient connection failures can be detected using [GRPC keepalives](https://pkg.go.dev/google.golang.org/grpc/keepalive) along with the client [`WaitForStateChange`](https://pkg.go.dev/google.golang.org/grpc#ClientConn.WaitForStateChange) API. The time it takes to detect a dead connection is determined by the keepalive `Timeout` parameter. The grpc client will automatically try to reconnect with an exponential backoff policy.
+
+For future backwards compatibility the proxy teleport version will be included in the grpc client/server headers. This will allow either a client or server to downgrade messages accordingly.
+
+Metrics will be added so we can monitor whether a single grpc connection becomes a bottleneck for many concurrent streams. The following metrics will be tracked:
+
+1. Total throughput in bytes, this is the aggregate number of bytes sent over all grpc channels on a single connection.
+2. Number of concurrent streams, this is the number of streams at any instant.
+3. Total number of streams, this is the total number of streams including both current and past streams.
+
+With these metrics we will be able to see if throughput begins to flatten as more streams are being used. If this does become an issue additional tcp connections will need to be created.
+
+### Reverse Tunnel Agent Pool
+Changes to the reverse tunnel agent and agent pool are required to support this design. The existing implementation creates a connection to every proxy. The new impelementation will decide how many connections to create dynamically based on the `ClusterNetworkingConfig`. If `proxy_peering` is enabled, a single connection will be created. If it is not enabled a connection to every proxy will be created. Old agents can continue connecting to every proxy regardless of the `ClusterNetworkingConfig`.
```

Comment:
Instead of only allowing a single connection from agents when `proxy_peering` is enabled, could the number of connections from agents be configurable? 

This has the benefit of better agent reachability during proxy upgrades and outages without having to do the `proxy initiate reconnects` flow described below which may difficult to integrate with for different load balancers and deployment types. It's also better than the current status quo where the agents need to connect to every proxy service instead of only a subset, allowing for better horizontal scalability of proxies.

As a caveat, allowing for multiple connections from agents would require that the `Agent Proxy Relationship` mapping described above be a one-to-many mapping instead of a one-to-one mapping.

Reply:
I like the idea of supporting multiple connections. This was something that was possible in an earlier iteration of this design. The only change I see required to support this now is storing a list of ProxyIDs in the `ServerV2Spec` rather than a single proxy id. The nonce/auth logic should still work.

With that, configuring multiple connections could be added in the future if not in the initial implementation.

`proxy initiate reconnects` shouldn't have any issues with load balancers as they will be sent over the existing ssh connection. Even with multiple connections I think there are still come benefit to having this. Eventually we may want to initiate reconnects periodically not just during a graceful shutdown.